OSDOCS-4047 and OSDOCS-3861: Operator workloads and PSA compliance (#57078)
openshift/openshift-docs: modules/osdk-ensuring-operator-workloads-run-restricted-psa.adoc

// Module included in the following assemblies:
//
// * operators/operator_sdk/osdk-complying-with-psa.adoc

:_content-type: PROCEDURE
[id="osdk-ensuring-operator-workloads-run-restricted-psa_{context}"]
= Ensuring Operator workloads run in namespaces set to the restricted pod security level

If you ensure your Operator workloads can run in namespaces set to the `restricted` pod security level, your Operator project can run on a wide variety of deployments and environments. 

[WARNING]
====

> * You must leave the `runAsUser` field empty. If your image requires a user, ensure the `userID` is defined in the image and not by the security context.

Author: Addressed in 6655a29

> * Use a workload resource, such as `Deployment` or `StatefulSet`, to manage your Operator's pod. If your Operator creates workload pods by using an account with escalated SCC privileges, the pods might fail.

Reviewer: (the reality is most operators are probably not creating pods directly anyway, so it's debatable how much we should get into this, and where.  I think it would be best to ensure we have really good documentation about how the PSA label syncer works, elsewhere in the documentation, including the implications on direct pod creation, and then refer people to that)

> * Use a workload resource, such as `Deployment` or `StatefulSet`, to manage your Operator's pod. If your Operator creates workload pods by using an account with escalated SCC privileges, the pods might fail.

Author: Addressed in 52d26a1

> * Use a workload resource, such as `Deployment` or `StatefulSet`, to manage your Operator's pod. If your Operator creates workload pods by using an account with escalated SCC privileges, the pods might fail.

Author: I created OSDOCS-5757 to track the follow up work. It is only a stub right now, but we can connect to write a better description and follow up tasks.

> * Use a workload resource, such as `Deployment` or `StatefulSet`, to manage your Operator's pod. If your Operator creates workload pods by using an account with escalated SCC privileges, the pods might fail.

Author: I removed the second warning. Please let me know if you think that is the right approach. As a rule of thumb, we are told that our downstream docs should attempt to cover 80% of use cases. I am not sure if that heuristic is helpful in this instance.

> * Use a workload resource, such as `Deployment` or `StatefulSet`, to manage your Operator's pod. If your Operator creates workload pods by using an account with escalated SCC privileges, the pods might fail.

Reviewer: i think that's fine

====

.Procedure

* To configure Operator workloads to run in namespaces set to the `restricted` pod security level, edit the namespace definition in your as shown in the following example:
+
.Example `config/manager/manager.yaml` file
[source,yaml]
----
...
spec:
 securityContext: <1>
   seccompProfile:
     type: RuntimeDefault
   runAsNonRoot: true
 containers:
   - name: <operator_workload_container>
     securityContext:
       allowPrivilegeEscalation: false
       capabilities:
         drop:
           - ALL
...
----

> <1> Leave the `seccompProfile` value unset to ensure your Operator project can run in {product-title} 4.10 and earlier. The security context constraint (SCC) defaults to the pod security profile of the namespace.

Reviewer: ```suggestion
<1> Setting this is recommended but is not compatible with {product-title} 4.10 and earlier.  Leave the `seccompProfile` value unset to ensure your Operator project can run in both older and newer versions of {product-title}. The security context constraint (SCC) defaults to the pod security profile of the namespace.
```

Author: Addressed in 6f6db0d

// Module included in the following assemblies:
//
// * operators/operator_sdk/osdk-complying-with-psa.adoc

:_content-type: PROCEDURE
[id="osdk-ensuring-operator-workloads-run-restricted-psa_{context}"]
= Ensuring Operator workloads run in namespaces set to the restricted pod security level

> If you ensure your Operator workloads can run in namespaces set to the `restricted` pod security level, your Operator project can run on a wide variety of deployments and environments.

Reviewer 2: This sentence sounds odd. Maybe lead with the want (Operator project can run...) before the how? 

>To ensure your Operator project can run on a wide variety of deployments and environments, configure the  Operator workloads to run in namespaces set to the `restricted` pod security level.

Author: Great suggestion. Thank you.

====

.Procedure

* To configure Operator workloads to run in namespaces set to the `restricted` pod security level, edit your Operator's namespace definition similar to the following examples:
+
[IMPORTANT]
====
It is recommended that you set the seccomp profile in your Operator's namespace definition. However, setting the seccomp profile is not supported in {product-title} 4.10 and earlier.
====

** For Operator projects that must run in only {product-title} 4.11 and later, edit your Operator's namespace definition similar to the following example:
+
.Example `config/manager/manager.yaml` file
[source,yaml]
----
...
spec:
 securityContext:
   seccompProfile:
     type: RuntimeDefault <1>
   runAsNonRoot: true
 containers:
   - name: <operator_workload_container>
     securityContext:
       allowPrivilegeEscalation: false
       capabilities:
         drop:
           - ALL
...
----
<1> By setting the seccomp profile type to `RuntimeDefault`, the SCC defaults to the pod security profile of the namespace.

> ** For Operator projects that must also run in {product-title} 4.10, edit your Operator's namespace definition similar to the following example:

Reviewer 2: > must also run in {product-title} 4.10

My ignorance of Operator projects (TIL!), you are saying that the projects must run in only 4.10, or in 4.10+?

Reviewer 2: Should this say 4.10 _**and earlier**_, as the call-out states?

Author: This one is tricky.

The first substep above describes what to do to follow the best practice, but it only works in 4.11+.

Operators are often developed to support multiple versions of OCP. So, if you want your project to run in 4.10+, you have to leave the field out your yaml file.

Author: I was working on this before 4.9 was EOL, so I think I can simplify this in both places.

> ** For Operator projects that must run in the currently supported  {product-title} versions, including 4.10 and earlier, edit your Operator's namespace definition similar to the following example:

Reviewer 3: I think 4.9 might already be out of support, so maybe the "and earlier" part isn't necessary anymore? And maybe could make it less complex, if it keeps enough of the same meaning?

(If you remove "and earlier" here, there are other places in the PR you could as well)

```suggestion
** For Operator projects that must also run in {product-title} 4.10, edit your Operator's namespace definition similar to the following example:
```

[WARNING]
====
* Use a workload resource, such as `Deployment` or `StatefulSet`, to manage your Operator's pod. If your Operator creates workload pods by using an account with escalated SCC privileges, the pods might fail.

> * You must leave the `runAsUser` field empty. If your image requires a user, it cannot be run under security context constraints (SCC) and restricted pod security enforcement.

Reviewer: > it cannot be run under security context constraints (SCC)

it cannot be run under restricted security context constraints (SCC)

Author: Addressed in 6946ad6

.Procedure

* To configure Operator workloads to run in namespaces set to the `restricted` pod security level, edit your Operator's namespace definition similar to the following examples:
+
[IMPORTANT]
====
It is recommended that you set the seccomp profile in your Operator's namespace definition. However, setting the seccomp profile is not supported in {product-title} 4.10 and earlier.
====

** For Operator projects that must run in the currently supported  {product-title} versions, including 4.10 and earlier, edit your Operator's namespace definition similar to the following example:
+
.Example `config/manager/manager.yaml` file
[source,yaml]
----
...
spec:
 securityContext:
   seccompProfile:

>      type: '' <1>

Reviewer: i don't think this is valid, was it tested?  i think you need to leave out the seccompProfile field entirely in order to leave it "empty"

Author: Addressed in 6946ad6. To make it easier to understand, I swapped the steps for >=4.11 and <=4.10. It was a little odd annotating the absence of a field. Please let me know if this seems legible to you.
